primitives/src/trie/trie_proof.rs

```diff
@@ -123,32 +139,72 @@ impl TrieProof {
         if !self.verify(root_hash) {
             return Err(Error("proof invalid"));
         }
-        let mut todo: BTreeSet<KeyNibbles> = keys.iter().cloned().collect();
+        // Sort the keys by their proving nodes so that we can go through the proof and the
+        // requested keys simultaneously.
+        let mut keys: Vec<_> = keys
+            .iter()
+            .cloned()
+            .map(|k| (self.missing_proven_by.get(k).unwrap_or(k), k))
+            .collect();
+        keys.sort_by(|&(k1p, k1), &(k2p, k2)| {
+            k1p.post_order_cmp(k2p).then_with(|| k1.post_order_cmp(k2))
+        });
+
+        let mut keys = keys.into_iter();
+        let mut nodes = self.nodes.into_iter();
+
+        let mut cur_key = keys.next();
+        let mut cur_node = nodes.next();
         let mut result = BTreeMap::new();
-        for node in self.nodes {
-            if todo.remove(&node.key) {
-                let key = node.key.clone();
-                match node.into_value() {
-                    None => return Err(Error("required value not present")),
-                    Some(value) => {
-                        assert!(result.insert(key, value).is_none());
+        loop {
+            let cmp = cur_key
+                .and_then(|(k, _)| cur_node.as_ref().map(|n| k.post_order_cmp(&n.key)))
+                .unwrap_or(Ordering::Equal);
+            match (cmp, cur_key, cur_node) {
+                (_, None, None) => break,
+                (_, Some(_), None) | (Ordering::Less, Some(_), Some(_)) => {
+                    return Err(Error("not all requested proof values are present"))
+                }
+                (_, None, Some(n)) | (Ordering::Greater, Some(_), Some(n)) => {
+                    if n.into_value().ok().and_then(|v| v).is_some() {
+                        return Err(Error("value present that wasn't requested"));
                     }
+                    cur_node = nodes.next();
                 }
-            } else {
-                // Nodes that weren't requested shouldn't have a value.
-                if node.into_value().is_some() {
-                    return Err(Error("value present that wasn't requested"));
+                (Ordering::Equal, Some((_, k)), Some(n)) => {
+                    if !n.key.is_prefix_of(k) {
+                        return Err(Error("tried to prove key with non-parent"));
+                    }
+                    if n.key == *k {
+                        let key = n.key.clone();
+                        let value = n
+                            .into_value()
+                            .map_err(|_| Error("requested hybrid node value missing"))?;
+                        assert!(result.insert(key, value).is_none());
+                        cur_key = keys.next();
+                        cur_node = nodes.next();
+                    } else {
+                        eprintln!("checking {} against {} for negativity", k, n.key);
```

Comment:
Still a remaining `eprintln`. Should be a `log::debug`?

Reply:
Hmmm hmmm, not sure how I forgot that one… Removed and added a comment instead.